### ISSUE TITLE
[BUG] Blob client freeze when downloading blob (full framework only)

### ISSUE THREAD

**Issue by Socolin** [Azure/azure-sdk-for-net#10027]
**Describe the bug**
The code hang on `BlobClient.DownloadAsync()`

**Expected behavior**
It should not "freeze"

**Actual behavior (include Exception or Stack Trace)**
It freeze and prevent code from executing what next :(

**To Reproduce**

1. Create Azure blob storage premium

![1](https://user-images.githubusercontent.com/539620/74773765-87fc0500-5260-11ea-967b-e2200e9431b8.png)

2. Create a container `test`
![2](https://user-images.githubusercontent.com/539620/74773796-9518f400-5260-11ea-8fe0-433e6d88c315.png)

3. Uploads 2 files names `test.txt` and `test2.txt` with content `abc`
![image](https://user-images.githubusercontent.com/539620/74773839-ac57e180-5260-11ea-97d7-1597a0c4214c.png)

4. Run the following code (don't forget to set the `ConnectionString`)

```
using System;
using System.Diagnostics;
using System.Net.Http;
using System.Threading.Tasks;
using Azure;
using Azure.Core.Pipeline;
using Azure.Storage.Blobs;

namespace ReproBugBlobHang
{
	internal static class Program
	{
		private const string ContainerName = "test";
		private const string ConnectionString = "";

		private static readonly string[] Urls =
		{
			"test.txt",
			"test2.txt",
		};

		private static void Main()
		{
			var blobClientOptions = new BlobClientOptions {Transport = new HttpClientTransport(new HttpClient() {Timeout = TimeSpan.FromSeconds(1)})};
			var t1 = RunAsync(blobClientOptions);
			var t2 = RunAsync(blobClientOptions);
			Task.WaitAll(t1, t2);
		}

		private static async Task RunAsync(BlobClientOptions blobClientOptions)
		{
			foreach (var url in Urls)
			{
				var stopwatch = Stopwatch.StartNew();
				try
				{
					var connection = new BlobClient(ConnectionString, ContainerName, url, blobClientOptions);
					Console.WriteLine("BEGIN");
					var response = await connection.DownloadAsync();
					Console.WriteLine("END");
					Console.WriteLine(response);
				}
				catch (RequestFailedException ex) when (ex.Status == 404)
				{
					Console.WriteLine("NOT FOUND");
				}

				Console.WriteLine(stopwatch.ElapsedMilliseconds);
			}
		}
	}
}
```
csproj
```
<Project Sdk="Microsoft.NET.Sdk">

	<PropertyGroup>
		<OutputType>Exe</OutputType>
		<TargetFramework>net472</TargetFramework>
	</PropertyGroup>

	<ItemGroup>
		<PackageReference Include="Azure.Storage.Blobs" Version="12.3.0" />
		<PackageReference Include="System.Net.Http" Version="4.3.4" />
	</ItemGroup>

</Project>
```

Output:
```
D:/dev/git/rider/repro-jetbrains/ConsoleApplication2/ReproBugBlobHang/bin/Debug/net472/ReproBugBlobHang.exe
BEGIN
BEGIN
END
Status: 206, Value: Azure.Storage.Blobs.Models.BlobDownloadInfo
372
BEGIN
END
Status: 206, Value: Azure.Storage.Blobs.Models.BlobDownloadInfo
229
BEGIN
```
Then it hang until timeouts and throw
```
Unhandled Exception: System.AggregateException: One or more errors occurred. ---> System.AggregateException: Retry failed after 6 tries. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.HttpClientTransport.<ProcessAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.RequestActivityPolicy.<ProcessNextAsync>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.RequestActivityPolicy.<ProcessAsync>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.BufferResponsePolicy.<ProcessAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.LoggingPolicy.<ProcessAsync>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.LoggingPolicy.<ProcessAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.<ProcessAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.<ProcessAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at Azure.Core.Pipeline.RetryPolicy.<ProcessAsync>d__11.MoveNext()
   --- End of inner exception stack trace ---
   at Azure.Core.Pipeline.RetryPolicy.<ProcessAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.<ProcessAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.<ProcessAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Storage.Blobs.BlobRestClient.Blob.<DownloadAsync>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Storage.Blobs.Specialized.BlobBaseClient.<StartDownloadAsync>d__45.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Storage.Blobs.Specialized.BlobBaseClient.<DownloadInternal>d__44.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Storage.Blobs.Specialized.BlobBaseClient.<DownloadAsync>d__43.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Storage.Blobs.Specialized.BlobBaseClient.<DownloadAsync>d__41.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Storage.Blobs.Specialized.BlobBaseClient.<DownloadAsync>d__39.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at ReproBugBlobHang.Program.<RunAsync>d__4.MoveNext() in Program.cs:line 39
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.WaitAll(Task[] tasks, Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.WaitAll(Task[] tasks, Int32 millisecondsTimeout)
   at System.Threading.Tasks.Task.WaitAll(Task[] tasks)
   at ReproBugBlobHang.Program.Main() in Program.cs:line 27

```

**Environment:**
 - Azure.Storage.Blobs 12.3.0
 - Windows 10 .Net Framework 4.7.2  / .Net Framework

**Note**
It's working fine with TargetFramework: netcoreapp3.1

**Comment by jsquire** [comment]
//cc: @tg-msft

**Comment by seanmcc-msft** [comment]
Hi @Socolin,

Couple things: 

`var connection = new BlobClient(ConnectionString, ContainerName, url, blobClientOptions);` 
With this constructor, the 3rd parameter should be just the name of your blob, not its URL.

`var blobClientOptions = new BlobClientOptions {Transport = new HttpClientTransport(new HttpClient() {Timeout = TimeSpan.FromSeconds(1)})};`  

Is setting a client timeout.  I recommend you remove this line, it is causing your requests to be cancelled, resulting in the 
`Unhandled Exception: System.AggregateException: One or more errors occurred. ---> System.AggregateException: Retry failed after 6 tries. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.`

-Sean

**Comment by Socolin** [comment]
> Hi @Socolin,
> 
> Couple things:
> 
> `var connection = new BlobClient(ConnectionString, ContainerName, url, blobClientOptions);`
> With this constructor, the 3rd parameter should be just the name of your blob, not its URL.
It's just bad naming after a day trying to figuring what the problem is... the values for these variables  are `test.txt` and `test2.txt` in the test

> `var blobClientOptions = new BlobClientOptions {Transport = new HttpClientTransport(new HttpClient() {Timeout = TimeSpan.FromSeconds(1)})};`
> 
> Is setting a client timeout. I recommend you remove this line, it is causing your requests to be cancelled, resulting in the
> `Unhandled Exception: System.AggregateException: One or more errors occurred. ---> System.AggregateException: Retry failed after 6 tries. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.`
I tried without, and it's just timimg out after 100sec (+ 6 retry so it take 5 minutes to display an error) I just put it here so you can view the problem quickly

**Comment by seanmcc-msft** [comment]
Hi @Socolin ,

I copied your repo code exactly, and was unable to reproduce on .NET Framework 4.7.2.  I created a container with two Block Blobs, both of who's content is "abc" using the Azure Storage Explorer.

I'm seeing `test.txt` and `test2.txt` both being downloaded twice on Fiddler, and am not experiencing the "freeze".

```
BEGIN
BEGIN
END
Status: 206, Value: Azure.Storage.Blobs.Models.BlobDownloadInfo
577
END
Status: 206, Value: Azure.Storage.Blobs.Models.BlobDownloadInfo
257
BEGIN
BEGIN
END
Status: 206, Value: Azure.Storage.Blobs.Models.BlobDownloadInfo
102
END
Status: 206, Value: Azure.Storage.Blobs.Models.BlobDownloadInfo
103
```

**Comment by Socolin** [comment]
Hello,

I just tried again and I have the exact same problem (I can send you a video if you want)
What can I do to help you reproduce the problem or understand what's wrong ?

Regards,

**Comment by seanmcc-msft** [comment]
Hi @Socolin ,

I believe the issue is something to do with your environment, not the SDK itself.  Are you using a proxy, or is your internet connection limited in some way?

-Sean

**Comment by Socolin** [comment]
> Hi @Socolin ,
> 
> I believe the issue is something to do with your environment, not the SDK itself. Are you using a proxy, or is your internet connection limited in some way?
> 
> -Sean

I don't think it's related to my internet connection because of the following points, but I'll try from my home, later, too just to be sure.
- I don't have any proxy, or any limitation on my internet connection.
- This problem was replicated on a webapp on azure (Windows azure webapp)
- When I switch to .NET Core I don't have the problem anymore.

Does the `bin/` folder built on my machine could help  ?

**Comment by christianrondeau** [comment]
Hi! I'm not sure why this issue was closed, we provided sample code, can reproduce in Microsoft Azure consistently and the same code fails in .NET 4.7.1 but not in .NET Core 3.1. I'm not sure what information was expected from us?

**Comment by tg-msft** [comment]
Sorry for closing the issue - we're trying to make our bot a little more helpful and it was overeager here.

Can you take a look at https://github.com/Azure/azure-sdk-for-net/issues/11503?  I think it's a duplicate of this issue and we're trying out a fix for it.  We can reopen this issue otherwise.

**Comment by christianrondeau** [comment]
Thanks @tg-msft for responding so quickly! This issue looks indeed very similar. We can try our repro after #12103 has been released (due to timing we had to port some code to .NET Core, which resolved the issue) or earlier if we have the opportunity to build azure sdk from source. You can leave that issue closed in the meantime, I will report back if the issue still shows up.
